src/v/cloud_topics/l1/tests/simple_metastore_test.cc

```diff
@@ -0,0 +1,338 @@
+        ASSERT_TRUE(offsets_res.has_value());
+        ASSERT_EQ(0_o, offsets_res->start_offset);
+        ASSERT_EQ(11_o, offsets_res->next_offset);
+    }
+
+    // Now add the object right at the end, where it should be.
+    auto ometa = om_builder(oid2, 100)
+                   .add(tid_a, 11_o, 20_o, 1000_t, 2000_t, 0, 99)
+                   .build();
+    auto add_res = m.add_objects(om_list_t::single(std::move(ometa))).get();
+    ASSERT_TRUE(add_res.has_value()) << int(add_res.error());
+
+    auto offsets_res
+      = m.get_offsets(model::topic_id_partition::from(tid_a)).get();
+    ASSERT_TRUE(offsets_res.has_value());
+    ASSERT_EQ(0_o, offsets_res->start_offset);
+    ASSERT_EQ(21_o, offsets_res->next_offset);
+}
+
+TEST(SimpleMetastoreTest, TestAddWithOverlap) {
+    simple_metastore m;
+    {
+        auto ometa = om_builder(oid1, 100)
+                       .add(tid_a, 0_o, 10_o, 1000_t, 2000_t, 0, 99)
+                       .build();
+        auto add_res = m.add_objects(
+                          chunked_vector<metastore::object_metadata>::single(
+                            std::move(ometa)))
+                         .get();
+        ASSERT_TRUE(add_res.has_value()) << int(add_res.error());
+    }
+
+    {
+        // Add another object just below where we expect it.
+        auto ometa = om_builder(oid2, 100)
+                       .add(tid_a, 10_o, 20_o, 1000_t, 2000_t, 0, 99)
+                       .build();
+        auto add_res = m.add_objects(
+                          chunked_vector<metastore::object_metadata>::single(
+                            std::move(ometa)))
+                         .get();
+        ASSERT_FALSE(add_res.has_value()) << int(add_res.error());
+        ASSERT_EQ(metastore::errc::invalid_request, add_res.error());
+    }
+    {
+        // Add another object that fully overlaps with what exists.
+        auto ometa = om_builder(oid2, 100)
+                       .add(tid_a, 0_o, 10_o, 1000_t, 2000_t, 0, 99)
+                       .build();
+        auto add_res = m.add_objects(om_list_t::single(std::move(ometa))).get();
+        ASSERT_FALSE(add_res.has_value()) << int(add_res.error());
+        ASSERT_EQ(metastore::errc::invalid_request, add_res.error());
+    }
+}
+
+TEST(SimpleMetastoreTest, TestAddPastBeginning) {
+    simple_metastore m;
+    // Add the first object so it doesn't start at 0.
+    auto ometa = om_builder(oid1, 100)
+                   .add(tid_a, 1_o, 10_o, 1000_t, 2000_t, 0, 99)
+                   .build();
+    auto add_res = m.add_objects(om_list_t::single(std::move(ometa))).get();
+    ASSERT_FALSE(add_res.has_value()) << int(add_res.error());
+    ASSERT_EQ(metastore::errc::invalid_request, add_res.error());
+}
+
+TEST(SimpleMetastoreTest, TestAddGetOffsetBasic) {
+    simple_metastore m;
+    om_list_t os;
+    os.emplace_back(om_builder(oid1, 100)
+                      .add(tid_a, 0_o, 10_o, 1000_t, 2000_t, 0, 99)
```

Comment:
consideration for future/other work: having simple named structs here helps make the tests more readable (don't spend time changing now)